safehttp/plugins/cors/cors.go

```diff
@@ -0,0 +1,208 @@
+)
+
+const customHeader string = "X-Cors"
+
+var disallowedMethods = map[string]bool{
+	safehttp.MethodHead: true,
+	safehttp.MethodGet:  true,
+	safehttp.MethodPost: true,
+}
+
+var disallowedContentTypes = map[string]bool{
+	"application/x-www-form-urlencoded": true,
+	"multipart/form-data":               true,
+	"text/plain":                        true,
+}
+
+// Interceptor handles CORS requests based on its settings.
+//
+// For more info about CORS, see: https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS
+//
+// Constraints
+//
+// - The request methods GET, HEAD and POST are banned and will result in a 405
+//   Method Not Allowed response.
+// - The content types "application/x-www-form-urlencoded", "multipart/form-data"
+//   and "text/plain" are also banned and will result in a  415 Unsupported Media
+//   Type response.
+// - Each CORS request must contain the header "X-Cors: 1" or "Sec-Fetch-Mode: cors".
+// All of this is to prevent XSRF.
+type Interceptor struct {
+	// AllowedOrigins determines which origins should be allowed in the
+	// Access-Control-Allow-Origin header.
+	AllowedOrigins map[string]bool
+	allowedHeaders map[string]bool
+	// ExposedHeaders determines which headers should be set in the
+	// Access-Control-Expose-Headers header. If ExposedHeaders is nil, then the
+	// header is not set.
+	ExposedHeaders []string
+	// AllowCredentials determines if Access-Control-Allow-Credentials should be
+	// set to true when requests containing Cookie headers arrive.
+	AllowCredentials bool
+	// MaxAge sets the Access-Control-Max-Age header.
+	// MaxAge=0 results in MaxAge: 5.
+	MaxAge int
+}
+
+// Default creates a CORS Interceptor with default settings.
+// Those defaults are:
+//   - No Exposed Headers
+//   - No Allowed Headers
+//   - AllowCredentials: false
+//   - MaxAge: 5 seconds
+func Default(allowedOrigins ...string) *Interceptor {
+	ao := map[string]bool{}
+	for _, o := range allowedOrigins {
+		ao[o] = true
+	}
+	return &Interceptor{
+		AllowedOrigins: ao,
+		allowedHeaders: map[string]bool{},
+	}
+}
+
+// SetAllowedHeaders sets which headers should be allowed in the Access-Control-Allow-Headers
+// header. The headers are first canonicalized using textproto.CanonicalMIMEHeaderKey.
+func (it *Interceptor) SetAllowedHeaders(headers ...string) {
+	it.allowedHeaders = map[string]bool{}
+	for _, h := range headers {
+		it.allowedHeaders[textproto.CanonicalMIMEHeaderKey(h)] = true
+	}
+}
+
+// Before handles the incoming request and sets headers or responds with an error
+// accordingly.
+func (it *Interceptor) Before(w safehttp.ResponseWriter, r *safehttp.IncomingRequest) safehttp.Result {
+	origin := r.Header.Get("Origin")
+	if !it.AllowedOrigins[origin] {
+		return w.ClientError(safehttp.StatusForbidden)
+	}
+	h := w.Header()
+	allowOrigin, err := h.Claim("Access-Control-Allow-Origin")
+	if err != nil {
+		return w.ServerError(safehttp.StatusInternalServerError)
+	}
+	vary, err := h.Claim("Vary")
```

Comment:
We will need `Vary` for many things, we cannot claim it. I think we might need to just set it (and preserve what was there already).

Reply:
Yup, I know. To be able to set it we need a `isClaimed` method on the `Header` type though. I talked about this with @kele and he said that we should just claim it for now to not add to much to the API and that other things that needed to set the Vary header should instead talk with the CORS plugin.

Reply:
Can't do that right now since there is no way of checking if the header is claimed without also setting it. :/ We would leak headers to error responses.

Reply:
Are we still blocked on this?

Reply:
No, so I will fix it